File: hnswlib-scala/src/main/scala/com/github/jelmerk/knn/scalalike/SparseVector.scala
package com.github.jelmerk.knn.scalalike

object SparseVector {

  /**
    * Extract fields from SparseVector.
    *
    * @param vector vector to extract the fields from
    * @tparam TVector type of values array
    * @return the extracted fields
    */
  def unapply[TVector](vector: SparseVector[TVector]): Option[(Array[Int], TVector)] =
    Some(vector.indices -> vector.values)

  /**
    * SparseVector factory.
    *
    * @param indices the index array, must be in ascending order
    * @param values the values array
    * @tparam TVector type of values array
    * @return a new SparseVector
    */
  def apply[TVector](indices: Array[Int], values: TVector): SparseVector[TVector] =
    new SparseVector[TVector](indices, values)

}

File: hnswlib-core/src/main/java/com/github/jelmerk/knn/SparseVector.java
package com.github.jelmerk.knn;

import java.io.Serializable;

/**
 * A sparse vector represented by an index array and a value array.
 */
public class SparseVector<TVector> implements Serializable {

    private static final long serialVersionUID = 1L;

    private int[] indices;
    private TVector values;

    /**
     * Constructs a new SparseVector instance.
     *
     * @param indices the index array, must be in ascending order
     * @param values the values array
     */
    public SparseVector(int[] indices, TVector values) {
        this.indices = indices;
        this.values = values;
    }
    /**
     * Returns the index array. Values are returned in ascending order.
     *
     * @return the index array
     */
    public int[] indices() {
        return indices;
    }

    /**
     * Returns the values array.
     *
     * @return the values array.
     */
    public TVector values() {
        return values;
    }
}
